Update development documentation for setup and workflow

GabLeRoux opened the pull request:
This pull request improves documentation for the GameCI Versioning Backend by restructuring the `DEVELOPMENT.md` file and expanding the `README.md` to provide a more comprehensive system overview. The changes aim to enhance clarity, usability, and accessibility for developers.

### Documentation Improvements:

#### Development Setup and Workflow
* [`DEVELOPMENT.md`](diffhunk://#diff-fc2a310fcfedfefb0046def697f932def80cbb1e78c689bc0af3c8eab3e33eceL1-R136): Restructured the document into clear sections, including prerequisites, setup instructions, credentials configuration, and troubleshooting tips. Added details on running Firebase locally, setting up environment variables, and deploying the backend.

#### System Overview and Features
* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L3-R79): Expanded to include a system overview with a mermaid diagram, detailed explanations of Unity version ingestion, CI job workflows, the scheduler, and the Ingeminator. Also added instructions for database backup and restoration.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Documentation**
  - Significantly expanded and restructured the development guide with clear setup steps, environment configuration, workflow instructions, and troubleshooting tips.
  - Enhanced the README with detailed system architecture, workflow descriptions, diagrams, and comprehensive operational procedures.
  - Added new sections for development and contribution guidelines to improve onboarding and reference for contributors.
- **Bug Fixes**
  - Updated emulator hosting port to avoid conflicts by changing it from 5000 to 5002.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

GabLeRoux committed:
Update development documentation for setup and workflow

GabLeRoux commented:
This is an improved version of #53 so I'll just close previous one.

NuclearCookie commented:
I'll post my feedback here as I follow the Development documentation:

When running `firebase emulators:start` I get the following log output:

```
⚠  emulators: You are not currently authenticated so some features may not work correctly. Please run firebase login to authenticate the CLI.
i  emulators: Shutting down emulators.

Error: Process `java -version` has exited with code 1. Please make sure Java is installed and on your system PATH.
-----Original stdout-----
-----Original stderr-----
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.
```

I'll now proceed with first logging in and making sure I have Java available in this environment, then continue to give feedback

GabLeRoux committed:
Update environment variable naming conventions in development documentation

NuclearCookie commented:
Logged in and made sure JAVA was available on path, now get this:
```
i  emulators: Starting emulators: functions, firestore, hosting, pubsub, storage, extensions
⬢  extensions: Unable to look up project number for unity-ci-versions.
 If this is a real project, ensure that you are logged in and have access to it.
 If this is a fake project, please use a project ID starting with 'demo-' to skip production calls.
 Continuing with a fake project number - secrets and other features that require production access may behave unexpectedly.
⚠  hosting: Port 5000 is not open on localhost (127.0.0.1,::1), could not start Hosting Emulator.
⚠  hosting: To select a different host/port, specify that host/port in a firebase.json config file:
      {
        // ...
        "emulators": {
          "hosting": {
            "host": "HOST",
            "port": "PORT"
          }
        }
      }
i  emulators: Shutting down emulators.

Error: Could not start Hosting Emulator, port taken.
```
This is due to Mac's port 5000 being used by airplay receiver. I'll continue with a different port and report back

NuclearCookie commented:
After changing the hosting port to 5002, the firebase command succeeded:

```

    "hosting": {
      "port": 5002
    },
    ```

GabLeRoux committed:
Correct spelling and update service account file path in database backup instructions

GabLeRoux committed:
Update service account file path in documentation

GabLeRoux committed:
Add Java Runtime Environment requirements and update emulator port configuration

NuclearCookie commented:
Now that I have the firebase CLI up and running, I do believe I have more issues.
Here's the output of running the emulators:

```
➜  versioning-backend git:(main) firebase emulators:start
i  emulators: Starting emulators: functions, firestore, hosting, pubsub, storage, extensions
⬢  extensions: Unable to look up project number for unity-ci-versions.
 If this is a real project, ensure that you are logged in and have access to it.
 If this is a fake project, please use a project ID starting with 'demo-' to skip production calls.
 Continuing with a fake project number - secrets and other features that require production access may behave unexpectedly.
⚠  functions: The following emulators are not running, calls to these services from the Functions emulator will affect production: apphosting, auth, database, dataconnect
⚠  functions: Unable to fetch project Admin SDK configuration, Admin SDK behavior in Cloud Functions emulator may be incorrect.
i  firestore: downloading cloud-firestore-emulator-v1.19.8.jar...
Progress: =============================================================================================================> (100% of 64MB)
i  firestore: Firestore Emulator logging to firestore-debug.log
✔  firestore: Firestore Emulator UI websocket is running on 9150.
i  pubsub: downloading pubsub-emulator-0.8.14.zip...
Progress: =============================================================================================================> (100% of 67MB)
i  pubsub: Pub/Sub Emulator logging to pubsub-debug.log
i  storage: downloading cloud-storage-rules-runtime-v1.1.3.jar...
Progress: =============================================================================================================> (100% of 53MB)
⚠  hosting: Authentication error when trying to fetch your current web app configuration, have you run firebase login?
⚠  hosting: Could not fetch web app configuration and there is no cached configuration on this machine. Check your internet connection and make sure you are authenticated. To continue, you must call firebase.initializeApp({...}) in your code before using Firebase.
i  hosting[unity-ci-versions]: Serving hosting files from: public
✔  hosting[unity-ci-versions]: Local server: http://127.0.0.1:5002
i  ui: downloading ui-v1.14.0.zip...
Progress: ==============================================================================================================> (100% of 4MB)
i  functions: Watching "/Users/pietervantorre/Projects/github/game-ci/versioning-backend/functions" for Cloud Functions...
⚠  functions: package.json indicates an outdated version of firebase-functions. Please upgrade using npm install --save firebase-functions@latest in your functions directory.
⬢  functions: Failed to load function definition from source: FirebaseError: There was an error reading functions/package.json:

 functions/lib/index.js does not exist, can't deploy Cloud Functions

┌─────────────────────────────────────────────────────────────┐
│ ✔  All emulators ready! It is now safe to connect your app. │
│ i  View Emulator UI at http://127.0.0.1:4000/               │
└─────────────────────────────────────────────────────────────┘

┌────────────┬────────────────┬──────────────────────────────────┐
│ Emulator   │ Host:Port      │ View in Emulator UI              │
├────────────┼────────────────┼──────────────────────────────────┤
│ Functions  │ 127.0.0.1:5001 │ http://127.0.0.1:4000/functions  │
├────────────┼────────────────┼──────────────────────────────────┤
│ Firestore  │ 127.0.0.1:8080 │ http://127.0.0.1:4000/firestore  │
├────────────┼────────────────┼──────────────────────────────────┤
│ Hosting    │ 127.0.0.1:5002 │ n/a                              │
├────────────┼────────────────┼──────────────────────────────────┤
│ Pub/Sub    │ 127.0.0.1:8085 │ n/a                              │
├────────────┼────────────────┼──────────────────────────────────┤
│ Storage    │ 127.0.0.1:9199 │ http://127.0.0.1:4000/storage    │
├────────────┼────────────────┼──────────────────────────────────┤
│ Extensions │ 127.0.0.1:5001 │ http://127.0.0.1:4000/extensions │
└────────────┴────────────────┴──────────────────────────────────┘
  Emulator Hub host: 127.0.0.1 port: 4400
  Other reserved ports: 4500, 9150
┌─────────────────────────┬───────────────┬─────────────────────┐
│ Extension Instance Name │ Extension Ref │ View in Emulator UI │
└─────────────────────────┴───────────────┴─────────────────────┘
Issues? Report them at https://github.com/firebase/firebase-tools/issues and attach the *-debug.log files
```

Most notably it can't deploy the cloud functions. 
Note that I didn't setup the credentials. Is that what I'm missing here?
It wasn't clear to me why I needed those

GabLeRoux committed:
Update DEVELOPMENT.md to clarify Firebase emulator port configuration and avoid conflicts

GabLeRoux commented:
> After changing the hosting port to 5002, the firebase command succeeded:
> 
> ```
> 
>     "hosting": {
>       "port": 5002
>     },
>     ```
> ```

I've applied your suggestion here: 09ab610 and here 0bf2450 to prevent conflict with Apple Airplay receiver on port 5000  on macOS :+1:

GabLeRoux committed:
Update DEVELOPMENT.md to include steps for building functions code and setting up credentials

GabLeRoux commented:
> Now that I have the firebase CLI up and running, I do believe I have more issues. Here's the output of running the emulators:
> 
> ```
> ➜  versioning-backend git:(main) firebase emulators:start
> i  emulators: Starting emulators: functions, firestore, hosting, pubsub, storage, extensions
> ⬢  extensions: Unable to look up project number for unity-ci-versions.
>  If this is a real project, ensure that you are logged in and have access to it.
>  If this is a fake project, please use a project ID starting with 'demo-' to skip production calls.
>  Continuing with a fake project number - secrets and other features that require production access may behave unexpectedly.
> ⚠  functions: The following emulators are not running, calls to these services from the Functions emulator will affect production: apphosting, auth, database, dataconnect
> ⚠  functions: Unable to fetch project Admin SDK configuration, Admin SDK behavior in Cloud Functions emulator may be incorrect.
> i  firestore: downloading cloud-firestore-emulator-v1.19.8.jar...
> Progress: =============================================================================================================> (100% of 64MB)
> i  firestore: Firestore Emulator logging to firestore-debug.log
> ✔  firestore: Firestore Emulator UI websocket is running on 9150.
> i  pubsub: downloading pubsub-emulator-0.8.14.zip...
> Progress: =============================================================================================================> (100% of 67MB)
> i  pubsub: Pub/Sub Emulator logging to pubsub-debug.log
> i  storage: downloading cloud-storage-rules-runtime-v1.1.3.jar...
> Progress: =============================================================================================================> (100% of 53MB)
> ⚠  hosting: Authentication error when trying to fetch your current web app configuration, have you run firebase login?
> ⚠  hosting: Could not fetch web app configuration and there is no cached configuration on this machine. Check your internet connection and make sure you are authenticated. To continue, you must call firebase.initializeApp({...}) in your code before using Firebase.
> i  hosting[unity-ci-versions]: Serving hosting files from: public
> ✔  hosting[unity-ci-versions]: Local server: http://127.0.0.1:5002
> i  ui: downloading ui-v1.14.0.zip...
> Progress: ==============================================================================================================> (100% of 4MB)
> i  functions: Watching "/Users/pietervantorre/Projects/github/game-ci/versioning-backend/functions" for Cloud Functions...
> ⚠  functions: package.json indicates an outdated version of firebase-functions. Please upgrade using npm install --save firebase-functions@latest in your functions directory.
> ⬢  functions: Failed to load function definition from source: FirebaseError: There was an error reading functions/package.json:
> 
>  functions/lib/index.js does not exist, can't deploy Cloud Functions
> 
> ┌─────────────────────────────────────────────────────────────┐
> │ ✔  All emulators ready! It is now safe to connect your app. │
> │ i  View Emulator UI at http://127.0.0.1:4000/               │
> └─────────────────────────────────────────────────────────────┘
> 
> ┌────────────┬────────────────┬──────────────────────────────────┐
> │ Emulator   │ Host:Port      │ View in Emulator UI              │
> ├────────────┼────────────────┼──────────────────────────────────┤
> │ Functions  │ 127.0.0.1:5001 │ http://127.0.0.1:4000/functions  │
> ├────────────┼────────────────┼──────────────────────────────────┤
> │ Firestore  │ 127.0.0.1:8080 │ http://127.0.0.1:4000/firestore  │
> ├────────────┼────────────────┼──────────────────────────────────┤
> │ Hosting    │ 127.0.0.1:5002 │ n/a                              │
> ├────────────┼────────────────┼──────────────────────────────────┤
> │ Pub/Sub    │ 127.0.0.1:8085 │ n/a                              │
> ├────────────┼────────────────┼──────────────────────────────────┤
> │ Storage    │ 127.0.0.1:9199 │ http://127.0.0.1:4000/storage    │
> ├────────────┼────────────────┼──────────────────────────────────┤
> │ Extensions │ 127.0.0.1:5001 │ http://127.0.0.1:4000/extensions │
> └────────────┴────────────────┴──────────────────────────────────┘
>   Emulator Hub host: 127.0.0.1 port: 4400
>   Other reserved ports: 4500, 9150
> ┌─────────────────────────┬───────────────┬─────────────────────┐
> │ Extension Instance Name │ Extension Ref │ View in Emulator UI │
> └─────────────────────────┴───────────────┴─────────────────────┘
> Issues? Report them at https://github.com/firebase/firebase-tools/issues and attach the *-debug.log files
> ```
> 
> Most notably it can't deploy the cloud functions. Note that I didn't setup the credentials. Is that what I'm missing here? It wasn't clear to me why I needed those

The part here:

>  functions/lib/index.js does not exist, can't deploy Cloud Functions

I think you need to build first (`cd functions && yarn build`), I've updated docs to reflect this. Also, yes, it's useful to create a firebase project and authenticate with it as well. I've updated the docs accordingly too: 3c36a64

NuclearCookie commented:
I confirm I got further after building! 
~Can I propose to maybe add that step explicitly to the setup documentation, and not just at the bottom for the Development Workflow? As far as I understood now it's an essential step to run when setting up the project for the first time.~
Edit: My bad, looked at the outdated documentation! LGTM

I will now try to create the firebase project and authenticate and report back

NuclearCookie commented:
I have setup the firebase project and authenticated, but don't see much of a difference.

Note that I also tried to make sure the env variable exists by trying like this:
`export GOOGLE_APPLICATION_CREDENTIALS="/Users/user/Projects/github/game-ci/versioning-backend/gameci-versioning-backend-test-firebase-adminsdk-fbsvc-b9db02a91a.json" && firebase emulators:start`

Here's the output from the command now:

```
➜  versioning-backend git:(main) ✗ firebase login
Already logged in as pieter@triangle-factory.be
➜  versioning-backend git:(main) ✗ export GOOGLE_APPLICATION_CREDENTIALS="/Users/pietervantorre/Projects/github/game-ci/versioning-backend/gameci-versioning-backend-test-firebase-adminsdk-fbsvc-b9db02a91a.json" && firebase emulators:start
i  emulators: Starting emulators: functions, firestore, hosting, pubsub, storage, extensions
⬢  extensions: Unable to look up project number for unity-ci-versions.
 If this is a real project, ensure that you are logged in and have access to it.
 If this is a fake project, please use a project ID starting with 'demo-' to skip production calls.
 Continuing with a fake project number - secrets and other features that require production access may behave unexpectedly.
⚠  functions: The following emulators are not running, calls to these services from the Functions emulator will affect production: apphosting, auth, database, dataconnect
⚠  functions: Application Default Credentials detected. Non-emulated services will access production using these credentials. Be careful!
⚠  functions: Unable to fetch project Admin SDK configuration, Admin SDK behavior in Cloud Functions emulator may be incorrect.
i  firestore: Firestore Emulator logging to firestore-debug.log
✔  firestore: Firestore Emulator UI websocket is running on 9150.
i  pubsub: Pub/Sub Emulator logging to pubsub-debug.log
⚠  hosting: Authentication error when trying to fetch your current web app configuration, have you run firebase login?
⚠  hosting: Could not fetch web app configuration and there is no cached configuration on this machine. Check your internet connection and make sure you are authenticated. To continue, you must call firebase.initializeApp({...}) in your code before using Firebase.
i  hosting[unity-ci-versions]: Serving hosting files from: public
✔  hosting[unity-ci-versions]: Local server: http://127.0.0.1:5002
i  functions: Watching "/Users/pietervantorre/Projects/github/game-ci/versioning-backend/functions" for Cloud Functions...
⚠  functions: package.json indicates an outdated version of firebase-functions. Please upgrade using npm install --save firebase-functions@latest in your functions directory.
⚠  functions: Your requested "node" version "20" doesn't match your global version "23". Using node@23 from host.
Serving at port 8421

webber.nl@gmail.com: Credential implementation provided to initializeApp() via the "credential" property failed to fetch a valid Google OAuth2 access token with the following error: "Error fetching access token: Error while making request: getaddrinfo ENOTFOUND metadata.google.internal. Error code: ENOTFOUND"..
lebreton.gabriel@gmail.com: Credential implementation provided to initializeApp() via the "credential" property failed to fetch a valid Google OAuth2 access token with the following error: "Error fetching access token: Error while making request: getaddrinfo ENOTFOUND metadata.google.internal. Error code: ENOTFOUND"..
davidmfinol@gmail.com: Credential implementation provided to initializeApp() via the "credential" property failed to fetch a valid Google OAuth2 access token with the following error: "Error fetching access token: Error while making request: getaddrinfo ENOTFOUND metadata.google.internal. Error code: ENOTFOUND"..
andrewk010110@gmail.com: Credential implementation provided to initializeApp() via the "credential" property failed to fetch a valid Google OAuth2 access token with the following error: "Error fetching access token: Error while making request: getaddrinfo ENOTFOUND metadata.google.internal. Error code: ENOTFOUND"..

⚠  functions: Application Default Credentials detected. Non-emulated services will access production using these credentials. Be careful!
✔  functions: Loaded functions definitions from source: model.editorVersion.onCreate, model.repoVersion.onCreate, cron.trigger, queueStatus, reportBuildFailure, reportNewBuild, reportPublication, unityVersions, retryBuild, testFunction.
✔  functions[europe-west3-model-editorVersion-onCreate]: firestore function initialized.
✔  functions[europe-west3-model-repoVersion-onCreate]: firestore function initialized.
✔  functions[europe-west3-cron-trigger]: pubsub function initialized.
✔  functions[europe-west3-queueStatus]: http function initialized (http://127.0.0.1:5001/unity-ci-versions/europe-west3/queueStatus).
✔  functions[europe-west3-reportBuildFailure]: http function initialized (http://127.0.0.1:5001/unity-ci-versions/europe-west3/reportBuildFailure).
✔  functions[europe-west3-reportNewBuild]: http function initialized (http://127.0.0.1:5001/unity-ci-versions/europe-west3/reportNewBuild).
✔  functions[europe-west3-reportPublication]: http function initialized (http://127.0.0.1:5001/unity-ci-versions/europe-west3/reportPublication).
✔  functions[europe-west3-unityVersions]: http function initialized (http://127.0.0.1:5001/unity-ci-versions/europe-west3/unityVersions).
✔  functions[europe-west3-retryBuild]: http function initialized (http://127.0.0.1:5001/unity-ci-versions/europe-west3/retryBuild).
✔  functions[europe-west3-testFunction]: http function initialized (http://127.0.0.1:5001/unity-ci-versions/europe-west3/testFunction).

┌─────────────────────────────────────────────────────────────┐
│ ✔  All emulators ready! It is now safe to connect your app. │
│ i  View Emulator UI at http://127.0.0.1:4000/               │
└─────────────────────────────────────────────────────────────┘

┌────────────┬────────────────┬──────────────────────────────────┐
│ Emulator   │ Host:Port      │ View in Emulator UI              │
├────────────┼────────────────┼──────────────────────────────────┤
│ Functions  │ 127.0.0.1:5001 │ http://127.0.0.1:4000/functions  │
├────────────┼────────────────┼──────────────────────────────────┤
│ Firestore  │ 127.0.0.1:8080 │ http://127.0.0.1:4000/firestore  │
├────────────┼────────────────┼──────────────────────────────────┤
│ Hosting    │ 127.0.0.1:5002 │ n/a                              │
├────────────┼────────────────┼──────────────────────────────────┤
│ Pub/Sub    │ 127.0.0.1:8085 │ n/a                              │
├────────────┼────────────────┼──────────────────────────────────┤
│ Storage    │ 127.0.0.1:9199 │ http://127.0.0.1:4000/storage    │
├────────────┼────────────────┼──────────────────────────────────┤
│ Extensions │ 127.0.0.1:5001 │ http://127.0.0.1:4000/extensions │
└────────────┴────────────────┴──────────────────────────────────┘
  Emulator Hub host: 127.0.0.1 port: 4400
  Other reserved ports: 4500, 9150
┌─────────────────────────┬───────────────┬─────────────────────┐
│ Extension Instance Name │ Extension Ref │ View in Emulator UI │
└─────────────────────────┴───────────────┴─────────────────────┘
Issues? Report them at https://github.com/firebase/firebase-tools/issues and attach the *-debug.log files
```

NuclearCookie commented:
I'm guessing my empty firebase project doesn't have the admin SDK setup yet. Trying to figure that out now:
https://firebase.google.com/docs/admin/setup

NuclearCookie commented:
Getting a little stuck here, I think I need to somehow have an existing backend as the firebase project. 
I'll wait for a response before continuing :) 
Great contributions so far, this will make it way easier to contribute in the future!
